packages/react-form-state/src/FormState.ts

```diff
@@ -186,52 +180,103 @@ export default class FormState<
+      onChange: this.updateField.bind(this, fieldPath),
+      onBlur: this.blurField.bind(this, fieldPath),
+    });
   }
-  private updateField<K extends keyof Fields>(fieldPath: K, value: Fields[K]) {
+
+  private updateField<Key extends keyof Fields>(
+    fieldPath: Key,
+    value: Fields[Key],
+  ) {
     this.setState<any>(({fields, dirtyFields}: State<Fields>) => {
       const field = fields[fieldPath];
-      const {error} = field;
-
       const dirty = !isEqual(value, field.initialValue);
-      const newDirtyFields = new Set(dirtyFields);
 
-      if (dirty) {
-        newDirtyFields.add(fieldPath);
-      } else {
-        newDirtyFields.delete(fieldPath);
-      }
-
-      const newFieldData = Object.assign({}, field, {
+      const updatedField = this.getUpdatedField({
+        fieldPath,
+        field,
         value,
         dirty,
       });
 
-      /*
-        We only want to update errors as the user types if they
-        already have an error.
-        https://polaris.shopify.com/patterns/error-messages#section-form-validation
-      */
-      const shouldUpdateError = error != null;
-
-      const newError = shouldUpdateError
-        ? this.validateFieldValue(fieldPath, newFieldData)
-        : // eslint-disable-next-line no-undefined
-          undefined;
-
       return {
-        dirtyFields: [...Array.from(newDirtyFields)],
-        fields: Object.assign({}, fields, {
-          [fieldPath]: Object.assign({}, newFieldData, {error: newError}),
+        dirtyFields: this.getUpdatedDirtyFields({
+          fieldPath,
+          dirty,
+          dirtyFields,
         }),
+        fields:
+          updatedField === field
+            ? fields
+            : Object.assign({}, fields, {
+                [fieldPath]: updatedField,
+              }),
       };
     });
   }
 
+  private getUpdatedDirtyFields<Key extends keyof Fields>({
+    fieldPath,
+    dirty,
+    dirtyFields,
+  }: {
+    fieldPath: Key;
+    dirty: boolean;
+    dirtyFields: (keyof Fields)[];
+  }) {
+    const dirtyFieldsSet = new Set(dirtyFields);
+
+    if (dirty) {
+      dirtyFieldsSet.add(fieldPath);
+    } else {
+      dirtyFieldsSet.delete(fieldPath);
+    }
+
+    const newDirtyFields = Array.from(dirtyFieldsSet);
+    return isEqual(dirtyFields, newDirtyFields) ? dirtyFields : newDirtyFields;
+  }
+
+  private getUpdatedField<Key extends keyof Fields>({
+    fieldPath,
+    field,
+    value,
+    dirty,
+  }: {
+    fieldPath: Key;
+    field: FieldStates<Fields>[Key];
+    value: Fields[Key];
+    dirty: boolean;
+  }) {
+    const newField = Object.assign({}, field, {
+      value,
+      dirty,
+    });
+
+    /*
+    We only want to update errors as the user types if they
+    already have an error.
+    https://polaris.shopify.com/patterns/error-messages#section-form-validation
+    */
+    const shouldUpdateError = field.error != null;
+    newField.error = shouldUpdateError
+      ? this.validateFieldValue(fieldPath, newField)
```

Comment:
Would it work with `newField.error = shouldUpdateError && this.validateFieldValue(fieldPath, newField)`?

Reply:
Not really sure why you are mutating the object instead of making this a property this in the third argument to `newField`?

Reply:
I don't think so because `shouldUpdateError` would be false not undefined.

Reply:
I actually don't remember why I mutated here, I'll take a look when I go over this this afternoon

Reply:
It was because of some awkwardness comparing the field object... Found a much nicer way to structure this though